File: app/src/main/java/com/davismiyashiro/expenses/view/managetabs/ChooseTabsActivity.java
package com.davismiyashiro.expenses.view.managetabs;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.davismiyashiro.expenses.Injection;
import com.davismiyashiro.expenses.R;
import com.davismiyashiro.expenses.view.addtab.AddTabActivity;
import com.davismiyashiro.expenses.model.BaseCompatActivity;
import com.davismiyashiro.expenses.datatypes.Tab;
import com.davismiyashiro.expenses.view.opentab.OpenTabActivity;

import java.util.ArrayList;
import java.util.List;

public class ChooseTabsActivity extends BaseCompatActivity implements ChooseTabsInterfaces.View {

    private FloatingActionButton fabAddTab;
    private ChooseTabsInterfaces.UserActionsListener mPresenter;
    private RecyclerView mTabsRecyclerView;
    private TabsAdapter mTabsAdapter;

    private LinearLayout mTabsView;
    private View mNoTabsView;
    private ImageView mNoTabsIcon;
    private TextView mNoTabsText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.tabs);
        super.onCreate(savedInstanceState);

//        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
//        setSupportActionBar(toolbar);

        mTabsRecyclerView = (RecyclerView) findViewById(R.id.tabs_recycler_view);
        mTabsRecyclerView.setLayoutManager(new
                GridLayoutManager(this, 2));
        mTabsAdapter = new TabsAdapter(new ArrayList<Tab>(), this);
        mTabsRecyclerView.setAdapter(mTabsAdapter);
        mTabsView = (LinearLayout) findViewById(R.id.tabs_root);

        mNoTabsView = findViewById(R.id.tabs_root_no_data);
        mNoTabsIcon = (ImageView) findViewById(R.id.tabs_no_data_icon);
        mNoTabsText = (TextView) findViewById(R.id.tabs_no_data_text_view);

        fabAddTab = (FloatingActionButton) findViewById(R.id.fab_manage_tabs);
        fabAddTab.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mPresenter.addTab();
                    }
                }
        );
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                //finish();
                break;
            default:
                // If we got here, the user's action was not recognized.
                // Invoke the superclass to handle it.
                return super.onOptionsItemSelected(item);
        }
        return true;
    }

    @Override
    public void onResume() {
        super.onResume();

        mPresenter = new ChooseTabsPresenterImpl(this, Injection.provideTabsRepository(getApplicationContext()));

        mPresenter.loadTabs(false);
    }

    @Override
    public void showAddTab() {
        //Intent intent = new Intent(this, AddTabActivity.class);
        startActivity(AddTabActivity.newIntent(this, null));
    }

    @Override
    public void setProgressIndicator(boolean active) {
    }

    @Override
    public void showTabs(List<Tab> tabs) {
        mTabsAdapter.replaceData(tabs);

        mTabsView.setVisibility(View.VISIBLE);
        mNoTabsView.setVisibility(View.GONE);
    }

    @Override
    public void showNoTabs() {
        loadShowNoTabsView(getString(R.string.tabs_no_data_text), R.drawable.ic_launcher);
    }

    private void loadShowNoTabsView (String message, int icon) {
        mTabsView.setVisibility(View.GONE);
        mNoTabsView.setVisibility(View.VISIBLE);

        mNoTabsText.setText(message);
        mNoTabsIcon.setImageDrawable(ContextCompat.getDrawable(this, icon));
    }

    @Override
    public void showTabDetailUi(Tab tab) {
        startActivity(OpenTabActivity.newInstance(this, tab));
    }

    /**
     * Receives tabs and bind them to the ViewHolder
     */
    private class TabsAdapter extends RecyclerView.Adapter<TabsAdapter.TabRowHolder> {

        private List<Tab> mTabs;
        private Context mContext;

        public TabsAdapter(List<Tab> tabs, Context context) {
            setList (tabs);
            mContext = context;
        }

        @Override
        public TabRowHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater layoutInflater =
                    LayoutInflater.from(getApplicationContext());
            View view = layoutInflater
                    .inflate(R.layout.list_item_tab, parent,
                            false);

            return new TabRowHolder(view);
        }

        @Override
        public void onBindViewHolder(TabRowHolder rowHolder, int position) {
            Tab tab = mTabs.get(position);
            //rowHolder.mTabTitleTextView.setText(tab.getGroupName());
            rowHolder.bindTab(tab);
        }

        public void replaceData(List<Tab> tabs) {
            setList(tabs);
            notifyDataSetChanged();
        }

        @Override
        public int getItemCount() {
            return mTabs.size();
        }

        private void setList(List<Tab> tabs) {
            mTabs = tabs;
        }

        public Tab getItem(int position) {
            return mTabs.get(position);
        }

        public void removeItemAtPosition(int position) {
            mTabs.remove(position);
            notifyItemRemoved(position);
        }

        public class TabRowHolder extends RecyclerView.ViewHolder {
            public TextView mTabTitleTextView;

            public TabRowHolder(View itemView) {
                super(itemView);
                mTabTitleTextView = (TextView) itemView.findViewById(R.id.list_item_tab_title);
                itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        int position = getAdapterPosition();
                        Tab tab = getItem(position);

                        mPresenter.openTab(tab.getGroupId());
                    }
                });

                itemView.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(mContext);
                        alertDialogBuilder.setTitle("Delete Tab");
                        alertDialogBuilder.setMessage("Are you sure you want to delete this Tab ?");

                        alertDialogBuilder.setPositiveButton("YES", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface arg0, int arg1) {
                                int position = getAdapterPosition();

                                mPresenter.removeTab(getItem(position).getGroupId());
                                removeItemAtPosition(position);
                                arg0.dismiss();
                            }
                        });

                        alertDialogBuilder.setNegativeButton("NO",new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.cancel();
                            }
                        });

//                        alertDialogBuilder.create().show();
                        AlertDialog alertDialog = alertDialogBuilder.create();
                        alertDialog.show();

                        return true;
                    }
                });
            }

            public void bindTab (Tab tab) {
                mTabTitleTextView.setText(tab.getGroupName());
            }
        }
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/view/managetabs/ChooseTabsActivity.kt
package com.davismiyashiro.expenses.view.managetabs

import android.content.Context
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.content.ContextCompat
import android.support.v7.app.AlertDialog
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.davismiyashiro.expenses.Injection
import com.davismiyashiro.expenses.R
import com.davismiyashiro.expenses.view.addtab.AddTabActivity
import com.davismiyashiro.expenses.view.BaseCompatActivity
import com.davismiyashiro.expenses.datatypes.Tab
import com.davismiyashiro.expenses.view.opentab.OpenTabActivity

import java.util.ArrayList

class ChooseTabsActivity : BaseCompatActivity(), ChooseTabsInterfaces.View {

    private lateinit var fabAddTab: FloatingActionButton
    private lateinit var mTabsRecyclerView: RecyclerView

    private lateinit var mTabsView: LinearLayout
    private lateinit var mNoTabsView: View
    private lateinit var mNoTabsIcon: ImageView
    private lateinit var mNoTabsText: TextView

    private lateinit var mTabsAdapter: TabsAdapter
    private var mPresenter: ChooseTabsInterfaces.UserActionsListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.tabs)
        super.onCreate(savedInstanceState)

        mTabsRecyclerView = findViewById(R.id.tabs_recycler_view)
        mTabsRecyclerView.layoutManager = GridLayoutManager(this, 2)
        mTabsAdapter = TabsAdapter(ArrayList(), this)
        mTabsRecyclerView.adapter = mTabsAdapter
        mTabsView = findViewById(R.id.tabs_root)

        mNoTabsView = findViewById(R.id.tabs_root_no_data)
        mNoTabsIcon = findViewById(R.id.tabs_no_data_icon)
        mNoTabsText = findViewById(R.id.tabs_no_data_text_view)

        fabAddTab = findViewById(R.id.fab_manage_tabs)
        fabAddTab.setOnClickListener { mPresenter?.addTab() }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onResume() {
        super.onResume()

        mPresenter = ChooseTabsPresenterImpl(this, Injection.provideTabsRepository(applicationContext))

        mPresenter?.loadTabs(false)
    }

    override fun showAddTab() {
        startActivity(AddTabActivity.newIntent(this, null))
    }

    override fun setProgressIndicator(active: Boolean) {}

    override fun showTabs(tabs: MutableList<Tab>) {
        mTabsAdapter.replaceData(tabs)

        mTabsView.visibility = View.VISIBLE
        mNoTabsView.visibility = View.GONE
    }

    override fun showNoTabs() {
        loadShowNoTabsView(getString(R.string.tabs_no_data_text), R.drawable.ic_launcher)
    }

    private fun loadShowNoTabsView(message: String, icon: Int) {
        mTabsView.visibility = View.GONE
        mNoTabsView.visibility = View.VISIBLE

        mNoTabsText.text = message
        mNoTabsIcon.setImageDrawable(ContextCompat.getDrawable(this, icon))
    }

    override fun showTabDetailUi(tab: Tab) {
        startActivity(OpenTabActivity.newInstance(this, tab))
    }

    /**
     * Receives tabs and bind them to the ViewHolder
     */
    private inner class TabsAdapter(tabs: MutableList<Tab>, private val mContext: Context) : RecyclerView.Adapter<TabsAdapter.TabRowHolder>() {

        private var mTabs = mutableListOf<Tab>()

        init {
            setList(tabs)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TabRowHolder {
            val layoutInflater = LayoutInflater.from(applicationContext)
            val view = layoutInflater
                    .inflate(R.layout.list_item_tab, parent,
                            false)

            return TabRowHolder(view)
        }

        override fun onBindViewHolder(rowHolder: TabRowHolder, position: Int) {
            val tab = mTabs[position]
            rowHolder.bindTab(tab)
        }

        fun replaceData(tabs: MutableList<Tab>) {
            setList(tabs)
            notifyDataSetChanged()
        }

        override fun getItemCount(): Int {
            return mTabs.size
        }

        private fun setList(tabs: MutableList<Tab>) {
            mTabs = tabs
        }

        fun getItem(position: Int): Tab {
            return mTabs[position]
        }

        fun removeItemAtPosition(position: Int) {
            mTabs.removeAt(position)
            notifyItemRemoved(position)
        }

        inner class TabRowHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            var mTabTitleTextView: TextView = itemView.findViewById(R.id.list_item_tab_title)

            init {
                itemView.setOnClickListener {
                    val position = adapterPosition
                    val (groupId) = getItem(position)

                    mPresenter?.openTab(groupId)
                }

                itemView.setOnLongClickListener { v ->
                    val alertDialogBuilder = AlertDialog.Builder(mContext)
                    alertDialogBuilder.setTitle("Delete Tab")
                    alertDialogBuilder.setMessage("Are you sure you want to delete this Tab ?")

                    alertDialogBuilder.setPositiveButton("YES") { arg0, _ ->
                        val position = adapterPosition

                        mPresenter?.removeTab(getItem(position).groupId)
                        removeItemAtPosition(position)
                        arg0.dismiss()
                    }

                    alertDialogBuilder.setNegativeButton("NO") { dialog, _ -> dialog.cancel() }
                    val alertDialog = alertDialogBuilder.create()
                    alertDialog.show()

                    true
                }
            }

            fun bindTab(tab: Tab) {
                mTabTitleTextView.text = tab.groupName
            }
        }
    }
}
